adapter/internal/discovery/xds/rate_limiter_cache.go

```diff
@@ -0,0 +1,253 @@
+					Key:       "method",
+					Value:     method,
+					RateLimit: rlPolicyConfig,
+				}
+				operationRlsConfigs = append(operationRlsConfigs, rlConf)
+				apiOperations[path][method] = void
+			}
+
+			if operationPolicyExists {
+				rlsConfig := &rls_config.RateLimitDescriptor{
+					Key:         "path",
+					Value:       path,
+					Descriptors: operationRlsConfigs,
+				}
+				rlsConfigs = append(rlsConfigs, rlsConfig)
+			}
+		}
+	} else {
+		return fmt.Errorf("invalid rate limit policy level: %q", level)
+	}
+
+	org := mgwSwagger.OrganizationID
+	vHost := mgwSwagger.VHost
+
+	if _, ok := r.apiLevelRateLimitPolicies[org]; !ok {
+		r.apiLevelRateLimitPolicies[org] = make(map[string]map[string][]*rls_config.RateLimitDescriptor)
+	}
+
+	if _, ok := r.apiLevelRateLimitPolicies[org][vHost]; !ok {
+		r.apiLevelRateLimitPolicies[org][vHost] = make(map[string][]*rls_config.RateLimitDescriptor)
+	}
+
+	r.apiLevelRateLimitPolicies[org][vHost][apiID] = rlsConfigs
+	return nil
+}
+
+// DeleteAPILevelRateLimitPolicies deletes inline Rate Limit policies added with the API.
+func (r *rateLimitPolicyCache) DeleteAPILevelRateLimitPolicies(org, vHost, apiID string) {
+	delete(r.apiLevelRateLimitPolicies[org][vHost], apiID)
+}
+
+func getRateLimitPolicy(policies map[string]*mgw.APIRateLimitPolicy, policyName string) (*rls_config.RateLimitPolicy, error) {
+	policy, ok := policies[policyName]
+	if !ok {
+		return nil, fmt.Errorf("rate limit policy %q not defined", policyName)
+	}
+
+	unit, err := getRateLimitUnit(policy.SpanUnit)
+	if err != nil {
+		return nil, err
+	}
+
+	return &rls_config.RateLimitPolicy{
+		Unit:            unit,
+		RequestsPerUnit: policy.Count,
+	}, nil
+}
+
+func (r *rateLimitPolicyCache) generateRateLimitConfig(label string) *rls_config.RateLimitConfig {
+	var orgDescriptors []*rls_config.RateLimitDescriptor
+
+	// Generate API level rate limit configurations
+	for org, orgPolicies := range r.apiLevelRateLimitPolicies {
+		var vHostDescriptors []*rls_config.RateLimitDescriptor
+		for vHost, vHostPolicies := range orgPolicies {
+			var apiDescriptors []*rls_config.RateLimitDescriptor
+			for apiID, apiPolicies := range vHostPolicies {
+				// Configure API Level rate limit policies only if, the API is deployed to the gateway label
+				// Check API deployed to the gateway label
+				if arrayContains(orgIDOpenAPIEnvoyMap[org][apiID], label) {
+					apiDescriptors = append(apiDescriptors, apiPolicies...)
+				}
+			}
+			vHostDescriptor := &rls_config.RateLimitDescriptor{
+				Key:         "vhost",
+				Value:       vHost,
+				Descriptors: apiDescriptors,
+			}
+			vHostDescriptors = append(vHostDescriptors, vHostDescriptor)
+		}
+		orgDescriptor := &rls_config.RateLimitDescriptor{
+			Key:         "org",
+			Value:       org,
+			Descriptors: vHostDescriptors,
+		}
+		orgDescriptors = append(orgDescriptors, orgDescriptor)
+	}
+
+	return &rls_config.RateLimitConfig{
+		Name:        "Default",
```

Comment:
IMO we can move the constants in routes_with_clusters file and reuse them instead of the hard-coded values. WDYT ?